### PR TITLE
[BEAM-7672] dynamically setup acceptable wheel specs according to installed python version

### PR DESCRIPTION
**Pull request by lazylynx**
Construct the accepted wheel based on the installed Python version.
This PR is the follow-up of #11970

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lazylynx** [comment]
R: @tvalentyn

**Comment by lazylynx** [comment]
@tvalentyn Cloud you re-trigger Python2_PVR_Flink?

**Comment by tvalentyn** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run Dataflow Python ValidatesContainer

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

File: sdks/python/container/boot.go [hunk 1]
```diff
@@ -170,6 +176,28 @@ func main() {
...
+func setupAcceptableWheelSpecs() error {
+	cmd := exec.Command("python", "-V")
+	stdoutStderr, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+	re := regexp.MustCompile(`Python (\d)\.(\d).*`)
+	pyVersions := re.FindStringSubmatch(string(stdoutStderr[:]))
+	if len(pyVersions) != 3 {
+		return fmt.Errorf("cannot get python version with expected format")
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:189]: How about we say "Cannot get parse Python version from ..." and append the output?

File: sdks/python/container/boot.go [hunk 2]
```diff
@@ -127,6 +129,10 @@ func main() {
 	// Guard from concurrent artifact retrieval and installation,
 	// when called by child processes in a worker pool.
 
+	if err := setupAcceptableWheelSpecs(); err != nil {
+		log.Fatalf("Failed to setup acceptable wheel specs: %v", err)
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:133]: I think we should not fail but leave acceptableWhlSpecs as empty in this case.

File: sdks/python/container/boot.go (hunk 1 above)
Comment (lazylynx) [reply sdks/python/container/boot.go:189]: It would be better!

File: sdks/python/container/boot.go (hunk 2 above)
Comment (lazylynx) [reply sdks/python/container/boot.go:133]: It's true. We should only output messages even when error returned.

**Comment by lazylynx** [comment]
@tvalentyn Updated. PTAL when tests are passed.

File: sdks/python/container/boot.go [hunk 3]
```diff
@@ -170,6 +176,28 @@ func main() {
...
+	re := regexp.MustCompile(`Python (\d)\.(\d).*`)
+	pyVersions := re.FindStringSubmatch(string(stdoutStderr[:]))
+	if len(pyVersions) != 3 {
+		return fmt.Errorf("cannot get parse Python version from %s", stdoutStderr)
+	}
+	pyVersion := fmt.Sprintf("%s%s", pyVersions[1], pyVersions[2])
+	if pyVersion == "27" {
+		acceptableWhlSpecs = append(acceptableWhlSpecs, "cp27-cp27mu-manylinux1_x86_64.whl")
+	} else {
+		wheelName := fmt.Sprintf("cp%s-cp%sm-manylinux1_x86_64.whl", pyVersion, pyVersion)
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:195]: Sorry for a slow response. It seems that 3.8 wheels don't follow this pattern, just noticing this while prepring 2.23.0 release. 3.8 wheels look as follows:

apache_beam-2.23.0-cp38-cp38-manylinux1_x86_64.whl

Not sure why the pattern has changed.

Comment (lazylynx) [reply sdks/python/container/boot.go:195]: Thank you for notification.
I found that wheel changed naming rule at 0.33.5.
> 0.33.5 (2019-08-17)
> Don’t add the m ABI flag to wheel names on Python 3.8 (PR by rdb)
https://wheel.readthedocs.io/en/stable/news.html

I'll push a fix tonight, wait a bit.

**Comment by lazylynx** [comment]
@tvalentyn Updated about wheel name with Python 3.8. PTAL.

File: sdks/python/container/boot.go [hunk 4]
```diff
@@ -170,6 +176,34 @@ func main() {
...
+		return err
+	}
+	re := regexp.MustCompile(`Python (\d)\.(\d).*`)
+	pyVersions := re.FindStringSubmatch(string(stdoutStderr[:]))
+	if len(pyVersions) != 3 {
+		return fmt.Errorf("cannot get parse Python version from %s", stdoutStderr)
+	}
+	pyVersion := fmt.Sprintf("%s%s", pyVersions[1], pyVersions[2])
+	if pyVersion == "27" {
+		acceptableWhlSpecs = append(acceptableWhlSpecs, "cp27-cp27mu-manylinux1_x86_64.whl")
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:193]: Can we move this into the `case` as well?  Thank you!

File: sdks/python/container/boot.go (hunk 3 above)
Comment (tvalentyn) [reply sdks/python/container/boot.go]: Thanks! Also filed https://issues.apache.org/jira/browse/BEAM-10434 to fix the stager.

File: sdks/python/container/boot.go (hunk 4 above)
Comment (lazylynx) [reply sdks/python/container/boot.go]: Updated, PTAL.

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Comment by tvalentyn** [comment]
Thanks, @lazylynx !

**Pull request merged**